src/Lecture2.hs

```diff
@@ -264,7 +345,30 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval _ (Lit value) = Right value
+eval variables (Var variable) = if value == Nothing
+                                then Left (VariableNotFound variable)
+                                else Right (fromJust value)
+  where value = lookup variable variables
+        fromJust               :: Maybe a -> a
+        fromJust (Just a) =  a
+        fromJust Nothing  =  error "fromJust: Nothing"
+
+eval variables (Add frst scnd) = if isRight evalFrst && isRight evalScnd
+                                 then Right (fromRight 0 evalFrst + fromRight 0 evalScnd)
+                                 else Left (pickLeft evalFrst evalScnd)
+  where evalFrst = eval variables frst
+        evalScnd = eval variables scnd
+        fromRight :: b -> Either a b -> b
+        fromRight _ (Right b) = b
+        fromRight b _         = b
+        isRight :: Either a b -> Bool
+        isRight (Left  _) = False
+        isRight (Right _) = True
+        pickLeft :: Either a b -> Either a b -> a
+        pickLeft (Left a) _ = a
+        pickLeft _ (Left a) = a
+        pickLeft _ _        = error "pickLeft: both Right"
```

Comment:
That's a great helper function! 👏🏻 
Once you learn about Monads, you'll be able to reduce this function to a single line ✂️